File: tools/scene_schema.py
"""What a scene block may say — checked before it reaches scenes.yaml.

One function, `validate(scene, zones=None) -> list[str]`: an empty list
means the scene is well-formed; otherwise every problem found, each as
one sentence the desk can show next to the field. The studio calls it
before splicing a block into the show (a 400 with the list, instead of a
clean write followed by a failed render) and gen_esphome calls it before
emitting, so the generator and the editor reject the same things.

The vocabularies come from gen_esphome — effect ids, overlays, palettes,
flash modes — rather than a second copy here: that module is the contract
with firmware/castle_effects.h, and a list kept in sync by comment is the
exact drift this file exists to prevent. Imported lazily because
gen_esphome imports this module.

Deliberately NOT checked: that `audio_file` exists (render_audio does,
against the configured library), and anything a synth or a pulse stream
interprets for itself — this is the shape of the block, not the show.
"""

from __future__ import annotations

import re
from typing import Any

import effect_vocab as ev

REQUIRED = ("id", "name", "kind", "duration_ms", "base")
CUE_OPS = ("set", "strike")
ID_RE = re.compile(r"^\w+$", re.ASCII)   # \w, but not the Unicode half


def _vocab() -> dict[str, set[str]]:
    return {"effect": set(ev.EFFECT_IDS), "overlay": set(ev.OVERLAY_IDS),
            "palette": set(ev.PALETTE_IDS), "pixels": set(ev.FLASH_MODE_IDS)}


def _num(v: Any) -> bool:
    return isinstance(v, (int, float)) and not isinstance(v, bool) \
        and v == v and v not in (float("inf"), float("-inf"))


def _unit(where: str, v: Any, errs: list[str], lo: float = 0.0,
          hi: float = 1.0) -> None:
    if not _num(v) or not lo <= v <= hi:
        errs.append(f"{where}: must be a number from {lo:g} to {hi:g}, got {v!r}")


def _color(where: str, v: Any, errs: list[str]) -> None:
    if (not isinstance(v, list) or len(v) not in (3, 4)
            or not all(_num(c) and 0 <= c <= 1 for c in v)):
        errs.append(f"{where}: must be [r, g, b(, w)] with each from 0 to 1")


def _zone(where: str, z: Any, zones: set[str] | None, errs: list[str]) -> None:
    if not isinstance(z, str) or not z:
        errs.append(f"{where}: zone must be a name")
    elif zones is not None and z not in zones:
        errs.append(f"{where}: no zone {z!r} (have {', '.join(sorted(zones))})")


def _effect(where: str, e: Any, vocab: dict[str, set[str]],
            errs: list[str]) -> None:
    if e not in vocab["effect"]:
        errs.append(f"{where}: unknown effect {e!r} "
                    f"(one of {', '.join(sorted(vocab['effect']))})")


def _in(where: str, v: Any, kind: str, vocab: dict[str, set[str]],
        errs: list[str]) -> None:
    if v not in vocab[kind]:
        errs.append(f"{where}: unknown {kind} {v!r} "
                    f"(one of {', '.join(sorted(vocab[kind]))})")


def _check_cue(i: int, c: Any, length: int | None, zones: set[str] | None,
               vocab: dict[str, set[str]], errs: list[str]) -> None:
    w = f"cues[{i}]"
    if not isinstance(c, dict):
        errs.append(f"{w}: must be a mapping")
        return
    t: Any = c.get("t")
    if not _num(t) or t < 0:
        errs.append(f"{w}: t must be a time in ms >= 0, got {t!r}")
    elif length is not None and t > length:
        errs.append(f"{w}: t={t:g} is past the scene's duration_ms={length}")
    op = c.get("op")
    if op not in CUE_OPS:
        errs.append(f"{w}: op must be one of {', '.join(CUE_OPS)}, got {op!r}")
        return
    if op == "set":
        if "zone" not in c:
            errs.append(f"{w}: a set cue needs a zone")
        else:
            _zone(f"{w}.zone", c["zone"], zones, errs)
        if "effect" not in c:
            errs.append(f"{w}: a set cue needs an effect")
        else:
            _effect(f"{w}.effect", c["effect"], vocab, errs)
        if "level" in c:
            _unit(f"{w}.level", c["level"], errs)
        return
    # strike
    if "zone" in c:
        _zone(f"{w}.zone", c["zone"], zones, errs)
    if "targets" in c:
        if not isinstance(c["targets"], list):
            errs.append(f"{w}.targets: must be a list of zones")
        else:
            for z in c["targets"]:
                _zone(f"{w}.targets", z, zones, errs)
    if "pixels" in c:
        _in(f"{w}.pixels", c["pixels"], "pixels", vocab, errs)
    if "intensity" in c:
        _unit(f"{w}.intensity", c["intensity"], errs, 0, 4)
    if "decay" in c:
        _unit(f"{w}.decay", c["decay"], errs)
    errs.extend(f"{w}.{k}: must be a number of ms >= 0, got {c[k]!r}"
                for k in ("ms", "attack")
                if k in c and (not _num(c[k]) or c[k] < 0))
    if "color" in c:
        _color(f"{w}.color", c["color"], errs)


def _pulse_zones(w: str, p: dict, zones: set[str] | None,
                 errs: list[str]) -> None:
    """Where a pulse lands: one zone, or lists of them."""
    if "zone" in p:
        _zone(f"{w}.zone", p["zone"], zones, errs)
    for k in ("zones", "boost_targets"):
        if k not in p:
            continue
        if not isinstance(p[k], list):
            errs.append(f"{w}.{k}: must be a list of zones")
            continue
        for z in p[k]:
            _zone(f"{w}.{k}", z, zones, errs)


def _pulse_colors(w: str, p: dict, errs: list[str]) -> None:
    """A pulse's colours: two named ones, plus an optional palette to pick
       from. A `colors:` that is present but empty is a mistake worth a
       message — the render would silently fall back to white."""
    for k in ("color", "color_hot"):
        if k in p:
            _color(f"{w}.{k}", p[k], errs)
    if "colors" not in p:
        return
    if not isinstance(p["colors"], list) or not p["colors"]:
        errs.append(f"{w}.colors: must be a non-empty list of colours")
        return
    for c in p["colors"]:
        _color(f"{w}.colors", c, errs)


def _pulse_shape(w: str, p: dict, vocab: dict[str, set[str]],
                 errs: list[str]) -> None:
    """How hard it hits and how it falls."""
    if "pixels" in p:
        _in(f"{w}.pixels", p["pixels"], "pixels", vocab, errs)
    if "intensity" in p:
        _unit(f"{w}.intensity", p["intensity"], errs, 0, 4)
    if "decay" in p:
        _unit(f"{w}.decay", p["decay"], errs)
    errs.extend(f"{w}.{k}: must be a number of ms >= 0, got {p[k]!r}"
                for k in ("ms", "attack_ms")
                if k in p and (not _num(p[k]) or p[k] < 0))


def _check_pulse(i: int, p: Any, zones: set[str] | None,
                 vocab: dict[str, set[str]], errs: list[str]) -> None:
    w = f"pulse[{i}]"
    if not isinstance(p, dict):
        errs.append(f"{w}: must be a mapping")
        return
    if not isinstance(p.get("synth"), str) or not p["synth"]:
        errs.append(f"{w}: needs a synth (the marker stream it follows)")
    _pulse_zones(w, p, zones, errs)
    _pulse_shape(w, p, vocab, errs)
    _pulse_colors(w, p, errs)


def _scene_head(scene: dict, errs: list[str]) -> int | None:
    """id, name, kind, duration, volume, loop, audio_file — the scalars.
       Returns the scene's length in ms when it has a usable one."""
    errs.extend(f"missing required key {k!r}" for k in REQUIRED if k not in scene)
    sid = scene.get("id")
    if "id" in scene and (not isinstance(sid, str) or not ID_RE.match(sid)):
        errs.append(f"id: letters, digits and _ only, got {sid!r}")
    errs.extend(f"{k}: must be a non-empty string" for k in ("name", "kind")
                if k in scene and (not isinstance(scene[k], str)
                                   or not scene[k].strip()))
    if "volume" in scene:
        _unit("volume", scene["volume"], errs)
    if "loop" in scene and not isinstance(scene["loop"], bool):
        errs.append(f"loop: must be true or false, got {scene['loop']!r}")
    if "audio_file" in scene:
        af = scene["audio_file"]
        if (not isinstance(af, str) or not af or af.startswith("/")
                or ".." in af.split("/")):
            errs.append(f"audio_file: must be a relative path, got {af!r}")
    if "duration_ms" not in scene:
        return None
    d = scene["duration_ms"]
    if not _num(d) or d <= 0 or int(d) != d:
        errs.append(f"duration_ms: must be a whole number of ms > 0, got {d!r}")
        return None
    return int(d)


def _mapping(name: str, value: Any, complaint: str, errs: list[str]) -> bool:
    """`value` is a mapping, or one error saying what it should have been."""
    if isinstance(value, dict):
        return True
    errs.append(f"{name}: {complaint}")
    return False


def _scene_zones(scene: dict, zs: set[str] | None,
                 vocab: dict[str, set[str]], errs: list[str]) -> None:
    """base, levels and zones: three maps keyed by zone, each with its own
       idea of what a value is."""
    if "base" in scene and _mapping(
            "base", scene["base"], "must map each zone to an effect", errs):
        for z, e in scene["base"].items():
            _zone("base", z, zs, errs)
            _effect(f"base.{z}", e, vocab, errs)
    levels = scene.get("levels")
    if levels is not None and _mapping(
            "levels", levels, "must map zones to a level", errs):
        for z, v in levels.items():
            _zone("levels", z, zs, errs)
            _unit(f"levels.{z}", v, errs)
    zd = scene.get("zones")
    if zd is not None and _mapping(
            "zones", zd, "must map zones to their texture", errs):
        for z, d in zd.items():
            _zone("zones", z, zs, errs)
            _zone_texture(z, d, vocab, errs)


def _zone_texture(z: str, d: Any, vocab: dict[str, set[str]],
                  errs: list[str]) -> None:
    """One zone's entry under `zones:` — centre effect, overlay, palette, phase."""
    if not _mapping(f"zones.{z}", d, "must be a mapping", errs):
        return
    if "center" in d:
        _effect(f"zones.{z}.center", d["center"], vocab, errs)
    for k in ("overlay", "palette"):
        if k in d:
            _in(f"zones.{z}.{k}", d[k], k, vocab, errs)
    if "phase" in d and not _num(d["phase"]):
        errs.append(f"zones.{z}.phase: must be a number")


def validate(scene: Any, zones: list[str] | None = None) -> list[str]:
    """Every way `scene` is not a scene — [] when it is one.

    `zones` is the show's zone list when the caller knows it; without it,
    zone names are only checked for shape.
    """
    if not isinstance(scene, dict):
        return ["scene must be a mapping"]
    errs: list[str] = []
    vocab = _vocab()
    zs = set(zones) if zones is not None else None
    length = _scene_head(scene, errs)
    _scene_zones(scene, zs, vocab, errs)
    cues = scene.get("cues")
    if cues is not None:
        if isinstance(cues, list):
            for i, c in enumerate(cues):
                _check_cue(i, c, length, zs, vocab, errs)
        else:
            errs.append("cues: must be a list")
    pulse = scene.get("pulse")
    if pulse is not None:
        if isinstance(pulse, list):
            for i, pl in enumerate(pulse):
                _check_pulse(i, pl, zs, vocab, errs)
        else:
            errs.append("pulse: must be a list")
    return errs

File: core/src/scene_schema.rs
//! What a scene block may say — `tools/scene_schema.py`, ported.
//!
//! One function, [`validate`]: an empty list means the scene is
//! well-formed; otherwise every problem found, each as one sentence the
//! desk shows next to the field. The studio calls it before splicing a
//! block into the show (a 400 with the list, instead of a clean write
//! followed by a failed render).
//!
//! There are two copies of these rules, and that is deliberate:
//! `tools/scene_schema.py` is what `gen_esphome.py` runs before it emits,
//! this is what the studio runs before it writes, and
//! `tests/test_scene_schema_rust.py` drives both over one corpus and
//! compares the sentences. The delegation through `tools/scene_check.py`
//! that used to make them one implementation went with the Python server
//! (docs/RETIREMENT.md phase 2); the parity gate is what replaced it.
//!
//! Deliberately NOT checked, exactly as in the Python: that `audio_file`
//! exists (render_audio does, against the configured library), and
//! anything a synth or a pulse stream interprets for itself — this is the
//! shape of the block, not the show.

use crate::scene_cues::{check_cue, check_pulse};
use crate::vocab;
use crate::yaml::{Yaml, format_g};

const REQUIRED: [&str; 5] = ["id", "name", "kind", "duration_ms", "base"];
pub(crate) const CUE_OPS: [&str; 2] = ["set", "strike"];

/// scene_schema.ID_RE — `^\w+$` with `re.ASCII`.
fn is_ident(s: &str) -> bool {
    !s.is_empty() && s.chars().all(|c| c.is_ascii_alphanumeric() || c == '_')
}

fn joined(names: &[&str]) -> String {
    let mut v: Vec<&str> = names.to_vec();
    v.sort_unstable();
    v.join(", ")
}

/// scene_schema._unit. `lo`/`hi` arrive as the text Python's `:g` prints.
pub(crate) fn unit(w: &str, v: Option<&Yaml>, errs: &mut Vec<String>, lo: f64, hi: f64) {
    let ok = v
        .and_then(Yaml::finite_num)
        .is_some_and(|n| lo <= n && n <= hi);
    if !ok {
        let got = v.map(Yaml::repr).unwrap_or_else(|| "None".into());
        errs.push(format!(
            "{w}: must be a number from {} to {}, got {got}",
            format_g(lo),
            format_g(hi)
        ));
    }
}

/// scene_schema._color.
pub(crate) fn color(w: &str, v: &Yaml, errs: &mut Vec<String>) {
    let ok = match v.as_list() {
        Some(items) if items.len() == 3 || items.len() == 4 => items
            .iter()
            .all(|c| c.finite_num().is_some_and(|n| (0.0..=1.0).contains(&n))),
        _ => false,
    };
    if !ok {
        errs.push(format!("{w}: must be [r, g, b(, w)] with each from 0 to 1"));
    }
}

/// scene_schema._zone. `zones` is None when the show's zone list is unknown.
pub(crate) fn zone(w: &str, z: &Yaml, zones: Option<&[String]>, errs: &mut Vec<String>) {
    match z.as_str() {
        None => errs.push(format!("{w}: zone must be a name")),
        Some("") => errs.push(format!("{w}: zone must be a name")),
        Some(name) => {
            if let Some(known) = zones {
                if !known.iter().any(|k| k == name) {
                    let mut have: Vec<&str> = known.iter().map(String::as_str).collect();
                    have.sort_unstable();
                    errs.push(format!(
                        "{w}: no zone {} (have {})",
                        z.repr(),
                        have.join(", ")
                    ));
                }
            }
        }
    }
}

/// scene_schema._zone for a mapping KEY, which is text by the time it is here.
fn zone_key(w: &str, k: &str, zones: Option<&[String]>, errs: &mut Vec<String>) {
    zone(w, &Yaml::Str(k.to_string()), zones, errs);
}

/// scene_schema._effect.
pub(crate) fn effect(w: &str, e: &Yaml, errs: &mut Vec<String>) {
    if !e.as_str().is_some_and(|s| vocab::EFFECTS.contains(&s)) {
        errs.push(format!(
            "{w}: unknown effect {} (one of {})",
            e.repr(),
            joined(&vocab::EFFECTS)
        ));
    }
}

/// scene_schema._in.
pub(crate) fn one_of(w: &str, v: &Yaml, kind: &str, names: &[&str], errs: &mut Vec<String>) {
    if !v.as_str().is_some_and(|s| names.contains(&s)) {
        errs.push(format!(
            "{w}: unknown {kind} {} (one of {})",
            v.repr(),
            joined(names)
        ));
    }
}

/// `ms` / `attack` / `attack_ms`: a number of milliseconds, never negative.
pub(crate) fn ms_fields(w: &str, m: &Yaml, keys: &[&str], errs: &mut Vec<String>) {
    for k in keys {
        if let Some(v) = m.get(k) {
            if !v.finite_num().is_some_and(|n| n >= 0.0) {
                errs.push(format!(
                    "{w}.{k}: must be a number of ms >= 0, got {}",
                    v.repr()
                ));
            }
        }
    }
}

/// scene_schema._scene_head — the scalars, and the scene's length in ms
/// when it has a usable one.
fn scene_head(scene: &Yaml, errs: &mut Vec<String>) -> Option<i64> {
    for k in REQUIRED {
        if !scene.has(k) {
            errs.push(format!("missing required key '{k}'"));
        }
    }
    if let Some(sid) = scene.get("id") {
        if !sid.as_str().is_some_and(is_ident) {
            errs.push(format!(
                "id: letters, digits and _ only, got {}",
                sid.repr()
            ));
        }
    }
    for k in ["name", "kind"] {
        if let Some(v) = scene.get(k) {
            if v.as_str().is_none_or(|s| s.trim().is_empty()) {
                errs.push(format!("{k}: must be a non-empty string"));
            }
        }
    }
    if scene.has("volume") {
        unit("volume", scene.get("volume"), errs, 0.0, 1.0);
    }
    if let Some(l) = scene.get("loop") {
        if !matches!(l, Yaml::Bool(_)) {
            errs.push(format!("loop: must be true or false, got {}", l.repr()));
        }
    }
    if let Some(af) = scene.get("audio_file") {
        let ok = af.as_str().is_some_and(|s| {
            !s.is_empty() && !s.starts_with('/') && !s.split('/').any(|p| p == "..")
        });
        if !ok {
            errs.push(format!(
                "audio_file: must be a relative path, got {}",
                af.repr()
            ));
        }
    }
    let d = scene.get("duration_ms")?;
    match d.finite_num() {
        Some(n) if n > 0.0 && n.fract() == 0.0 => Some(n as i64),
        _ => {
            errs.push(format!(
                "duration_ms: must be a whole number of ms > 0, got {}",
                d.repr()
            ));
            None
        }
    }
}

/// scene_schema._mapping — `value` is a mapping, or one error saying what
/// it should have been.
fn mapping<'a>(
    name: &str,
    v: &'a Yaml,
    complaint: &str,
    errs: &mut Vec<String>,
) -> Option<&'a [(String, Yaml)]> {
    match v.as_map() {
        Some(kv) => Some(kv),
        None => {
            errs.push(format!("{name}: {complaint}"));
            None
        }
    }
}

/// scene_schema._zone_texture — one zone's entry under `zones:`.
fn zone_texture(z: &str, d: &Yaml, errs: &mut Vec<String>) {
    if mapping(&format!("zones.{z}"), d, "must be a mapping", errs).is_none() {
        return;
    }
    if let Some(c) = d.get("center") {
        effect(&format!("zones.{z}.center"), c, errs);
    }
    for k in ["overlay", "palette"] {
        if let Some(v) = d.get(k) {
            let names: &[&str] = if k == "overlay" {
                &vocab::OVERLAYS
            } else {
                &vocab::PALETTES
            };
            one_of(&format!("zones.{z}.{k}"), v, k, names, errs);
        }
    }
    if let Some(p) = d.get("phase") {
        if p.finite_num().is_none() {
            errs.push(format!("zones.{z}.phase: must be a number"));
        }
    }
}

/// scene_schema._scene_zones — base, levels and zones.
fn scene_zones(scene: &Yaml, zs: Option<&[String]>, errs: &mut Vec<String>) {
    if let Some(base) = scene.get("base") {
        if let Some(kv) = mapping("base", base, "must map each zone to an effect", errs) {
            for (z, e) in kv {
                zone_key("base", z, zs, errs);
                effect(&format!("base.{z}"), e, errs);
            }
        }
    }
    if let Some(levels) = scene.get("levels").filter(|v| **v != Yaml::Null) {
        if let Some(kv) = mapping("levels", levels, "must map zones to a level", errs) {
            for (z, v) in kv {
                zone_key("levels", z, zs, errs);
                unit(&format!("levels.{z}"), Some(v), errs, 0.0, 1.0);
            }
        }
    }
    if let Some(zd) = scene.get("zones").filter(|v| **v != Yaml::Null) {
        if let Some(kv) = mapping("zones", zd, "must map zones to their texture", errs) {
            for (z, d) in kv {
                zone_key("zones", z, zs, errs);
                zone_texture(z, d, errs);
            }
        }
    }
}

/// Every way `scene` is not a scene — empty when it is one.
///
/// `zones` is the show's zone list when the caller knows it; without it,
/// zone names are only checked for shape.
pub fn validate(scene: &Yaml, zones: Option<&[String]>) -> Vec<String> {
    if scene.as_map().is_none() {
        return vec!["scene must be a mapping".into()];
    }
    let mut errs = Vec::new();
    let length = scene_head(scene, &mut errs);
    scene_zones(scene, zones, &mut errs);
    if let Some(cues) = scene.get("cues").filter(|v| **v != Yaml::Null) {
        match cues.as_list() {
            Some(items) => {
                for (i, c) in items.iter().enumerate() {
                    check_cue(i, c, length, zones, &mut errs);
                }
            }
            None => errs.push("cues: must be a list".into()),
        }
    }
    if let Some(pulse) = scene.get("pulse").filter(|v| **v != Yaml::Null) {
        match pulse.as_list() {
            Some(items) => {
                for (i, p) in items.iter().enumerate() {
                    check_pulse(i, p, zones, &mut errs);
                }
            }
            None => errs.push("pulse: must be a list".into()),
        }
    }
    errs
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::yaml::parse;

    const ZONES: [&str; 3] = ["towerL", "towerR", "door"];

    fn errs(text: &str) -> Vec<String> {
        let zones: Vec<String> = ZONES.iter().map(|z| (*z).to_string()).collect();
        validate(&parse(text).expect("parses"), Some(&zones))
    }

    /// Every problem, in the order the fields are read — the desk shows
    /// the first in its headline and the rest beneath it. The corpus-wide
    /// gates are tests/test_scene_schema_rust.py (against the Python) and
    /// tests/golden/scene_errors.json (against the recorded answers); this
    /// is the shape of one answer, readable in the crate.
    #[test]
    fn a_scene_full_of_mistakes_names_each_one_in_reading_order() {
        assert_eq!(
            errs(
                "id: trial\nkind: seance\nvolume: 4\nduration_ms: -3\n\
                  base: {moat: disco}\ncues: [{t: 5, op: wobble}]\n"
            ),
            vec![
                "missing required key 'name'",
                "volume: must be a number from 0 to 1, got 4",
                "duration_ms: must be a whole number of ms > 0, got -3",
                "base: no zone 'moat' (have door, towerL, towerR)",
                "base.moat: unknown effect 'disco' (one of blood, candle, chill, \
                 ember, eyes, furnace, mansion, off, seance, spirit, strobe, throb, wisp)",
                "cues[0]: op must be one of set, strike, got 'wobble'",
            ]
        );
        assert_eq!(
            validate(&parse("- a\n").expect("parses"), None),
            ["scene must be a mapping"]
        );
    }

    /// A minimal scene is clean, and a cue exactly at the end is inside it.
    #[test]
    fn a_well_formed_scene_says_nothing() {
        assert_eq!(
            errs(
                "id: probe\nname: Probe\nkind: triggered\nduration_ms: 5000\n\
                  base: {towerL: candle, towerR: candle, door: ember}\n\
                  cues: [{t: 5000, op: strike, ms: 70, pixels: scatter}]\n"
            ),
            Vec::<String>::new()
        );
    }
}
